src/main/scala/io/github/spark_redshift_community/spark/redshift/v2/RedshiftPreProcessor.scala

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.spark_redshift_community.spark.redshift.v2
+
+import java.io.InputStreamReader
+import java.net.URI
+
+import com.amazonaws.auth.AWSCredentialsProvider
+import com.amazonaws.services.s3.AmazonS3Client
+import com.eclipsesource.json.Json
+import io.github.spark_redshift_community.spark.redshift.{AWSCredentialsUtils, DefaultJDBCWrapper, FilterPushdown, JDBCWrapper, Utils}
+import io.github.spark_redshift_community.spark.redshift.Parameters.MergedParameters
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.SparkSession
+
+import scala.collection.JavaConverters._
+
+class RedshiftPreProcessor(schemaOpt: Option[StructType],
```

Comment:
The behaviour is similar(in the sense that the unload happens during planning as part of `io.github.spark_redshift_community.spark.redshift.RedshiftRelation#buildScan`) to dsv1.
Since this is not a distributed operation, this has to happen in the driver. In this case, I have tried to push operation to the end of the planning phase. Let me know your thoughts